File: 560.subarray-sum-equals-k.cpp
// @leet start
#include <vector>
#include <unordered_map>
using std::vector;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        // TC: O(n), SC: O(n)
        int total = 0;
        std::unordered_map<int, int> map;

        int prefixSum = 0;
        ++map[0];
        for (const int& ele : nums) {
            prefixSum += ele;
            total += map[prefixSum - k];
            ++map[prefixSum];
        }

        return total;
    }
};
// @leet end

File: 560.subarray-sum-equals-k.py
# @leet imports start
from collections import defaultdict
from typing import List, Optional

# @leet imports end

# @leet start
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        psums = defaultdict(int)
        psums[0] = 1
        ans = 0
        psum = 0
        for x in nums:
            psum += x
            ans += psums[psum - k]
            psums[psum] += 1

        return ans


# @leet end

n = [1, 1, 1]
k = 2
x = Solution().subarraySum(n, k)
print(x)

[1, 2, 3]
3

